### PR TITLE
Add an isometric grid option

### PR DESCRIPTION
**Pull request by D8H**
When it's enable it will a draw diamonds instead of rectangles for each cell.
It uses generators to draw the lines. The code is a lot more simple that way but it may not be a good idea on a draw function.

IsometricGridConf

This is not just graphical because the magnet logic also changes.

IsometricGridLogic

### CODE REVIEW

File: Core/GDCore/IDE/Dialogs/LayoutEditorCanvas/LayoutEditorCanvasOptions.cpp [hunk 1]
```diff
@@ -17,6 +17,7 @@ LayoutEditorCanvasOptions::LayoutEditorCanvasOptions()
       gridHeight(32),
       gridOffsetX(0),
       gridOffsetY(0),
+      gridType("retangular"),
```
Comment (Bouh) [inline Core/GDCore/IDE/Dialogs/LayoutEditorCanvas/LayoutEditorCanvasOptions.cpp:20]:
```suggestion
      gridType("rectangular"),
```

Comment (D8H) [reply Core/GDCore/IDE/Dialogs/LayoutEditorCanvas/LayoutEditorCanvasOptions.cpp]: Nice catch

**Comment by 4ian** [comment]
Interesting! Do you think it would be easy to make the isometry angle customizable (like in the top down behavior)?

I'm a bit concerned about using generators as they can end up generating a bunch of boilerplate code and nothing would really beat the performance of a simple function call.
How would you do it otherwise?

**Comment by D8H** [comment]
> Interesting! Do you think it would be easy to make the isometry angle customizable (like in the top down behavior)?

* For pixel isometry, it's easier to set the grid with width and height.
* For true isometry, I'm not sure there is a consensus.
   * In my case, I used a 64x64 square and its projected dimensions are on the screenshot. So, it would be easier to input isoSize=64 and angle=30.
   * But I imagine using a projected width that is an integer could be a good solution to make assets. Then, it will be width=90 angle=30.
   * Could do the same with height instead of width.

I think it would make the UI confusing. We may assume people doing true isometry know what they get themselves into and let them copy/paste float values. However, I think a way to apply it globally to every scene or related layouts could help a lot.

> I'm a bit concerned about using generators as they can end up generating a bunch of boilerplate code and nothing would really beat the performance of a simple function call.
> How would you do it otherwise?

It can be done with classic loops. It will be more difficult to read but probably a lot faster.

**Comment by D8H** [comment]
By the way, it's the same for path finding setting. I chose to use the width and height as dimension in the isometric world, but for the pixel isometry at least it must be the dimension after the projection.

The same logic must be used for settings in the grid and path finding so users can make it matched.

File: newIDE/app/src/Utils/GridHelpers.js [hunk 2]
```diff
@@ -1,12 +1,61 @@
...
+    if (gridWidth <= 0) {
+      pos[0] = Math.round(pos[0]);
+    } else {
+      pos[0] =
+        Math.floor((pos[0] - gridOffsetX) / gridWidth) * gridWidth +
+        gridOffsetX;
+    }
+
+    if (gridHeight <= 0) {
+      pos[1] = Math.round([1]);
```
Comment (4ian) [inline newIDE/app/src/Utils/GridHelpers.js:54]: `pos[1]` - I'm surprised this was not caught by flow type checking?

File: newIDE/app/src/Utils/GridHelpers.js [hunk 3]
```diff
@@ -1,12 +1,61 @@
...
+  gridOffsetX: number,
+  gridOffsetY: number,
+  gridType: number
 ) => {
-  if (gridSize <= 0) {
-    return Math.round(pos);
+  if (gridType === 'isometric') {
+    if (gridWidth <= 0 || gridHeight <= 0) {
+      pos[0] = Math.round(pos[0]);
+      pos[1] = Math.round([1]);
```
Comment (4ian) [inline newIDE/app/src/Utils/GridHelpers.js:14]: pos[1]?

File: newIDE/app/src/Utils/GridHelpers.js [hunk 4]
```diff
@@ -1,12 +1,61 @@
...
+      return;
+    }
+    // Why do we need this?
+    // Take a 2x2 squares, put a diamond inside each square, there is a 5th diamond in the center
+    let cellX = Math.round(((pos[0] - gridOffsetX) * 2) / gridWidth);
+    let cellY = Math.round(((pos[1] - gridOffsetY) * 2) / gridHeight);
+
+    if ((((cellX + cellY) % 2) + 2) % 2 == 1) {
```
Comment (4ian) [inline newIDE/app/src/Utils/GridHelpers.js:22]: Isn't this equivalent to `(cellX + cellY) % 2 == 1`?

File: newIDE/app/src/Utils/GridHelpers.js [hunk 5]
```diff
@@ -1,12 +1,61 @@
...
-  if (gridSize <= 0) {
-    return Math.round(pos);
+  if (gridType === 'isometric') {
+    if (gridWidth <= 0 || gridHeight <= 0) {
+      pos[0] = Math.round(pos[0]);
+      pos[1] = Math.round([1]);
+      return;
+    }
+    // Why do we need this?
+    // Take a 2x2 squares, put a diamond inside each square, there is a 5th diamond in the center
```
Comment (4ian) [inline newIDE/app/src/Utils/GridHelpers.js:18]: This comment is not really explaining the calculation, could you add something explaining what cellX/Y represents?

File: newIDE/app/src/SceneEditor/SetupGridDialog.js [hunk 6]
```diff
@@ -69,7 +71,7 @@ export default class SetupGridDialog extends Component {
               type="number"
               value={this.state.gridWidth}
               onChange={(e, value) =>
-                this.setState({ gridWidth: parseInt(value, 10) })
+                this.setState({ gridWidth: parseFloat(value, 10) })
```
Comment (4ian) [inline newIDE/app/src/SceneEditor/SetupGridDialog.js:74]: parseFloat is taking just the value argument, you can remove 10 which is unused.

File: newIDE/app/src/Utils/GridHelpers.js (hunk 4 above)
Comment (D8H) [reply newIDE/app/src/Utils/GridHelpers.js:22]: I think -3 % 2 == -1

**Review by 4ian — APPROVED** [review]
Looks great! Thanks for the additional comments, very useful 👍

**Pull request merged**